File: ml/features.py
"""Landmark -> feature vector.

This is the contract between training and the browser. `web/features.js` is a
line-for-line port and `tests/test_feature_parity.py` fails if the two ever
disagree, because a feature bug that only exists on one side of the wire is
invisible to a Python-only test suite.

The transform exists to remove everything about a hand that is not its shape:
where it is in frame, how big it is, how it is rotated, and which hand it is.
What survives is the pose, which is the only thing a letter is.
"""

import numpy as np

WRIST = 0
MIDDLE_MCP = 9

NUM_LANDMARKS = 21
FEATURE_DIM = NUM_LANDMARKS * 3


def normalize(landmarks, handedness="Right"):
    """landmarks: (21, 3) array in MediaPipe image coordinates.

    Returns a (63,) float32 feature vector.
    """
    pts = np.asarray(landmarks, dtype=np.float64).reshape(NUM_LANDMARKS, 3).copy()

    # Left hands are mirrored onto the right-hand manifold so that one model
    # serves both. MediaPipe's x axis runs left-to-right in image space, so a
    # sign flip is the whole mirror.
    if str(handedness).lower().startswith("l"):
        pts[:, 0] = -pts[:, 0]

    # 1. Translate: the wrist becomes the origin, removing position in frame.
    pts -= pts[WRIST]

    # 2. Rotate in the image plane so the wrist -> middle-finger-MCP axis points
    #    along -y ("up"). Removes how the wrist happened to be tilted.
    vx, vy = pts[MIDDLE_MCP, 0], pts[MIDDLE_MCP, 1]
    norm = np.hypot(vx, vy)
    if norm > 1e-8:
        cos_t, sin_t = -vy / norm, -vx / norm
        x, y = pts[:, 0].copy(), pts[:, 1].copy()
        pts[:, 0] = cos_t * x - sin_t * y
        pts[:, 1] = sin_t * x + cos_t * y

    # 3. Scale: farthest landmark from the wrist becomes unit distance, removing
    #    how close the hand was to the camera.
    scale = np.max(np.linalg.norm(pts, axis=1))
    if scale > 1e-8:
        pts /= scale

    return pts.reshape(-1).astype(np.float32)

File: web/features.js
// Port of ml/features.py. Kept deliberately literal so the two can be read
// side by side; tests/test_feature_parity.py runs this file under node and
// fails if it disagrees with the Python original on random inputs.

export const NUM_LANDMARKS = 21;
export const FEATURE_DIM = NUM_LANDMARKS * 3;

const WRIST = 0;
const MIDDLE_MCP = 9;

export function normalize(landmarks, handedness = "Right") {
  // landmarks: array of 21 {x, y, z} or a flat array of 63 numbers.
  // Three input shapes are accepted: MediaPipe's array of {x, y, z}, an array
  // of [x, y, z] triples, and a flat array of 63 numbers.
  const pts = new Float64Array(FEATURE_DIM);
  const head = landmarks[0];
  if (Array.isArray(head)) {
    for (let i = 0; i < NUM_LANDMARKS; i++) {
      pts[i * 3] = landmarks[i][0];
      pts[i * 3 + 1] = landmarks[i][1];
      pts[i * 3 + 2] = landmarks[i][2];
    }
  } else if (head !== null && typeof head === "object") {
    for (let i = 0; i < NUM_LANDMARKS; i++) {
      pts[i * 3] = landmarks[i].x;
      pts[i * 3 + 1] = landmarks[i].y;
      pts[i * 3 + 2] = landmarks[i].z;
    }
  } else {
    for (let i = 0; i < FEATURE_DIM; i++) pts[i] = landmarks[i];
  }

  if (String(handedness).toLowerCase().startsWith("l")) {
    for (let i = 0; i < NUM_LANDMARKS; i++) pts[i * 3] = -pts[i * 3];
  }

  const ox = pts[WRIST * 3], oy = pts[WRIST * 3 + 1], oz = pts[WRIST * 3 + 2];
  for (let i = 0; i < NUM_LANDMARKS; i++) {
    pts[i * 3] -= ox;
    pts[i * 3 + 1] -= oy;
    pts[i * 3 + 2] -= oz;
  }

  const vx = pts[MIDDLE_MCP * 3], vy = pts[MIDDLE_MCP * 3 + 1];
  const norm = Math.hypot(vx, vy);
  if (norm > 1e-8) {
    const cosT = -vy / norm, sinT = -vx / norm;
    for (let i = 0; i < NUM_LANDMARKS; i++) {
      const x = pts[i * 3], y = pts[i * 3 + 1];
      pts[i * 3] = cosT * x - sinT * y;
      pts[i * 3 + 1] = sinT * x + cosT * y;
    }
  }

  let scale = 0;
  for (let i = 0; i < NUM_LANDMARKS; i++) {
    const d = Math.hypot(pts[i * 3], pts[i * 3 + 1], pts[i * 3 + 2]);
    if (d > scale) scale = d;
  }
  if (scale > 1e-8) {
    for (let i = 0; i < FEATURE_DIM; i++) pts[i] /= scale;
  }

  return Float32Array.from(pts);
}
